dev/PipsControl/PipsControl.cpp

```diff
@@ -0,0 +1,369 @@
+    bool hiddenOnEdgeCondition,
+    ButtonVisibility visibility,
+    const wstring_view visibleStateName,
+    const wstring_view hiddenStateName,
+    const wstring_view enabledStateName,
+    const wstring_view disabledStateName) {
+
+    if (visibility != ButtonVisibility::Collapsed) {
+        if ((visibility == ButtonVisibility::Visible || m_isPointerOver) && !hiddenOnEdgeCondition) {
+            winrt::VisualStateManager::GoToState(*this, visibleStateName, false);
+            winrt::VisualStateManager::GoToState(*this, enabledStateName, false);
+        }
+        else {
+            if (hiddenOnEdgeCondition) {
+                winrt::VisualStateManager::GoToState(*this, disabledStateName, false);
+            }
+            winrt::VisualStateManager::GoToState(*this, hiddenStateName, false);
+        }
+    }
+}
+
+void PipsControl::OnNavigationButtonVisibilityChanged(ButtonVisibility visibility, const wstring_view collapsedStateName) {
+    if (visibility == ButtonVisibility::Collapsed) {
+        winrt::VisualStateManager::GoToState(*this, collapsedStateName, false);
+    }
+    else {
+        UpdateNavigationButtonVisualStates();
+    }
+}
+
+void PipsControl::UpdateNavigationButtonVisualStates() {
+
+    const int selectedPageIndex = SelectedPageIndex();
+    const int numberOfPages = NumberOfPages();
+    const int maxDisplayedPages = MaxDisplayedPages();
+    
+    if (!numberOfPages == 0 && maxDisplayedPages > 0) { 
+        auto const ifPreviousButtonHiddenOnEdge = selectedPageIndex == 0;
+        UpdateIndividualNavigationButtonVisualState(ifPreviousButtonHiddenOnEdge, PreviousButtonVisibility(),
+                c_previousPageButtonVisibleVisualState, c_previousPageButtonHiddenVisualState,
+                c_previousPageButtonEnabledVisualState, c_previousPageButtonDisabledVisualState);
+
+        auto const ifNextButtonHiddenOnEdge = selectedPageIndex == numberOfPages - 1;
+        UpdateIndividualNavigationButtonVisualState(ifNextButtonHiddenOnEdge, NextButtonVisibility(),
+            c_nextPageButtonVisibleVisualState, c_nextPageButtonHiddenVisualState,
+            c_nextPageButtonEnabledVisualState, c_nextPageButtonDisabledVisualState);
+    }
+}
+
+void PipsControl::OnElementPrepared(winrt::ItemsRepeater sender, winrt::ItemsRepeaterElementPreparedEventArgs args)
+{
+    if (const auto pip = args.Element().try_as<winrt::Button>()) {
+        auto const pageNumber = unbox_value<int>(pip.Tag());
+        if (unbox_value<int>(pip.Tag()) - 1 != SelectedPageIndex()) {
+            pip.Style(DefaultIndicatorStyle());
+        }
+        // Narrator says: Page 5, Button 5 of 30. Is it expected behavior?
+        winrt::AutomationProperties::SetName(pip, ResourceAccessor::GetLocalizedStringResource(SR_PipsControlPageText) + L" " + winrt::to_hstring(pageNumber));
+        winrt::AutomationProperties::SetPositionInSet(pip, pageNumber);
+        winrt::AutomationProperties::SetSizeOfSet(pip, NumberOfPages());
```

Comment:
When the size of the set changes you need to set the new size of set on all of the already prepared elements. You might also need to update the position in set if items were removed or added to the front of the collection. I think RadioButtons handles this if you want to take a look at what it does